Expand formalism docs

aphyr opened the pull request:
This rewrites and expands the developer platform formalism docs up to the storage & compute layer discussion. It begins with a quick sketch of the entire document, introduces new terms and properties for implicit concepts, adds a diagram to help clarify the way that frontiers bound pTVC times, uses partial orders for times throughout, decouples time-function from update-set representations of TVCs and partial TVCs, discusses compacting and appending to pTVCs and gestures vaguely in the direction of those transformations' correctness, and makes immutability/mutability and modeled vs real time explicit.

### Motivation

This PR aims to help make the formalism docs more approachable for new readers.

### Tips for reviewer

Several things in this rewrite I've guessed at or come up with names from scratch, so I'd appreciate detailed review!

See the rich markdown diff view!

### Testing

There are no code changes required

### Release notes

There are no user-facing changes, other than documentation.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MaterializeInc/materialize?pullRequest=11084) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MaterializeInc/materialize?pullRequest=11084) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MaterializeInc/materialize?pullRequest=11084) it.</sub>

frankmcsherry commented on doc/developer/platform/formalism.md:
```diff
@@ -13,57 +13,370 @@ It does not intend to describe the **current behavior** of the system.
+
+### pTVC append
+
+Let `ptvc = [ptvc-updates, since, upper]` be a correct view of some TVC's
+updates `tvc_updates`. Now consider appending a window of new updates to this
+pTVC, and advancing `upper` to `upper'`: `ptvc' = append(ptvc, upper',
+new-updates)`.
+
```
I may have missed it, in which case sorry, but it seems like this is crying out for a (notation failure) dTVC that is the differential of the TVC, which is all updates past and future, at which point a "correct pTVC in update form" has some relation to the dTVC (NB: not "the updates geq since and not geq upper" unfortunately). pTVC append should just be "add in those new updates from the dTVC", which we get to but with hand waving.

Will ponder if this could be more immediately obvious, with tweaked pre-work on the dTVC concept.

frankmcsherry replied:
I think if you have a dTVC, then a "correct pTVC in update form" has as updates the union of "the updates of the dTVC geq since and not geq upper" and "sufficient updates to cause the accumulations at each time in since to equal the TVC at that time". If you believed this, then it is probably clearer that pTVC append is "easy".

No opinion on whether this is helpful pedagogically, though.

aphyr replied:
> a ... dTVC that is the differential of the TVC

This is what we defined in "Update-set representations of TVCs and pTVCs". We call this the "update-set representation" of a TVC, as opposed to the "time-function representation". I didn't want to call it dTVC because we *also* then have a dpTVC. Instead, I opted to show an isomorphism between update-set and time-function representations, which lets us just talk about "a TVC" while using whichever representation is convenient. :)

A picture might help here--I'll draw one.

> pTVC append should just be "add in those new updates from the dTVC", which we get to but with hand waving.

That's exactly what's happening here. :-)

> I think if you have a dTVC, then a "correct pTVC in update form" has as updates the union of "the updates of the dTVC geq since and not geq upper" and "sufficient updates to cause the accumulations at each time in since to equal the TVC at that time". If you believed this, then it is probably clearer that pTVC append is "easy".

That's exactly what this section does! :D

frankmcsherry replied:
This is probably "I came to the same conclusion by a different path" and wanted to tell you about the path I took. No reason to fix anything here.

frankmcsherry commented on doc/developer/platform/formalism.md:
```diff
@@ -13,57 +13,370 @@ It does not intend to describe the **current behavior** of the system.
+## The TVC Map
+
+To track these changing pTVCs over time, Materialize maintains a mutable *TVC
+map*. The keys in this map are `GlobalId`s, and the values are pTVCs. Each TVC
+has at most one `GlobalID` over the entire history of a Materialize cluster. In
```
I think "Each TVC has at most one `GlobalId` ..." might want to be "Each `GlobalId` has at most one TVC ..."? What do you think?

aphyr replied:
I honestly don't know--this one's up to you! Is the goal to have multiple concurrent pTVCs, each with their own GlobalId, which correspond to a single TVC? I think this probably connects with whether you want the adapter's "source" to refer to TVCs or pTVCs too.

frankmcsherry replied:
I guess I'm reading "Each TVC has at most one `GlobalId`" to imply by omission that "a `GlobalId` can have multiple TVCs", which wouldn't make sense to me.

We can chew on this. My mental model is:
1. Materialize models changing data through an assignment to each `GlobalId` of *at most one* TVC.
2. The realization of that model assigns, at any moment, to each `GlobalId` a pTVC that is always compatible with the modeled TVC.

My hunch is that the same words mean different things to us, or that we are focusing on different implications of them, rather than that we disagree. I think.

aphyr replied:
I'm not disagreeing, haha--I literally don't know how Materialize is supposed to work, and I can see lots of defensible options. Let me try rewriting this based on your comment, and see what you think.

frankmcsherry reviewed (changes requested):
I put "request changes" because I think there are some points to discuss, but the high-level bit is that this is great and the world would be net better with it merged. But let's discuss the details first!

frankmcsherry commented on doc/developer/platform/formalism.md:
```diff
@@ -13,57 +13,370 @@ It does not intend to describe the **current behavior** of the system.
+time, Materialize uses sets of *updates* as a differential representation of a
+pTVC. These updates are triples of `[data, time, diff]`: the piece of data
+being changed, the time of the change, and the change in multiplicity of that
+data. Materialize can *compact* a pTVC by merging updates from the past, and
+*append* new updates to a pTVC as they arrive. Compaction advances the lower
+*since frontier* of the pTVC, and appending advances the *upper frontier*. Both
+compact and append preserve correctness: these differential representations of
+a pTVC are faithful to the "real" TVC between `since` and `upper`.
```
Actually, *compact* and *append* might be fine names for the non-differential form as well.

frankmcsherry commented on doc/developer/platform/formalism.md:
```diff
@@ -13,57 +13,370 @@ It does not intend to describe the **current behavior** of the system.
+
+A *frontier* is a set of times in `T`. We use frontiers to specify a lower (or
+upper) bound on times in `T`. We say that a time `t` *is later than* or *after*
+a frontier `F` iff there exists some `f` in `F` such that `f <t t`. We write
```
I think not, and retract this proposal in favor of the interpretation that `F <= t` should be `F <t t` instead.

frankmcsherry commented on doc/developer/platform/formalism.md:
```diff
@@ -13,57 +13,370 @@ It does not intend to describe the **current behavior** of the system.
+upper) bound on times in `T`. We say that a time `t` *is later than* or *after*
+a frontier `F` iff there exists some `f` in `F` such that `f <t t`. We write
+this `F <=t t`. Similarly we can speak of a time `t` which is later than or
+equal to F, by which we mean there exists some `f` in `F` such that `f <=t t`:
```
Ah, sorry I mean that the F (without ticks) should be an `F`.

aphyr replied:
Ah, got it!

frankmcsherry commented on doc/developer/platform/formalism.md:
```diff
@@ -13,57 +13,370 @@ It does not intend to describe the **current behavior** of the system.
+compact and append preserve correctness: these differential representations of
+a pTVC are faithful to the "real" TVC between `since` and `upper`.
+
+Materialize identifies each TVC by a unique `GlobalId`, and maintains a mutable map of `GlobalId`s to the current pTVC representing that `GlobalID`'s TVC.
```
> A GlobalID identifies a single TVC for all time (is this true? I think your comment above might mean... no?)

I meant it to be a single TVC for all time. I didn't realize I implied otherwise, so I might be miscommunicating.

The next two point (single pTVC and no first-class representation) are both spot on!

In MZ there is a `GlobalId`. That id is used in a few ways. For example, the catalog will store things like
```
CREATE VIEW foo AS
SELECT * FROM bar
```
using ids for `foo` and for `bar`. These correspond, imo, to the TVCs. We don't have the TVC at hand, but we are naming them and their relationships.

When you ship a dataflow to COMPUTE, you use the ids (say for `foo` and `bar` as in the query) to name the things you want to relate. At any moment one can resolve a pTVC associated with each of them, but their meaning is imo actually about the TVCs. E.g. at any moment maybe the pTVC for `foo` isn't yet caught up to `bar`. But it should be, because what we actually have is a relationships between TVCs, more than specific pTVCs.

The TVC is never reified in memory, but it is a real concept that some parts of the system mean to speak about (rather than speak about the associated pTVCs).

aphyr replied:
Thank you! This all makes sense. I'm going to add a section to the formalism that talks about GlobalIds specifically, and brings in some of these concepts, just in case they're not clear in the rest of the formalism.

frankmcsherry commented on doc/developer/platform/formalism.md:
```diff
@@ -13,57 +13,370 @@ It does not intend to describe the **current behavior** of the system.
+A *read* of a TVC `tvc` at time `t` means the collection version stored in
+`tvc` for `t`. We write this as `read(tvc, t) => collection-version`.
+
+Every TVC begins with the empty collection version: `read(tvc, t0) = cv0`.
```
I think it would line up more with how the underlying math and system actually works. I'm not sure if I convinced you later on, but I think it *avoids* a special case, because we already have an identity element in the accumulation function, and you just end up using this identity when the differencing devolves to "nothing prior to have accumulated".

frankmcsherry commented on doc/developer/platform/formalism.md:
```diff
@@ -13,57 +13,370 @@ It does not intend to describe the **current behavior** of the system.
+## Partial time-varying collections
+
+At any wall-clock time a TVC is (in general) unknowable. Some of its past
+changes have been forgotten, some information has yet to arrive, and its future
```
No strong feels about this. I thought it might be useful color, but I think I need an emoji to indicate :bonus: or something like that. 🍖 is the closest github offers me.

frankmcsherry commented on doc/developer/platform/formalism.md:
```diff
@@ -13,57 +13,370 @@ It does not intend to describe the **current behavior** of the system.
+At times between `since` and `upper`, a pTVC should be identical to its
+corresponding TVC.
+
+We think of `since` as the "read frontier": times not later than `since` cannot
```
No worries. I have the dual problem that I can no longer keep my mouth shut when I see glitches like this.

frankmcsherry commented on doc/developer/platform/formalism.md:
````diff
@@ -13,57 +13,370 @@ It does not intend to describe the **current behavior** of the system.
+                    }
+```
+
+Given a frontier `f`, a set of updates `updates` can be *compacted* together
+into a (likely smaller) set of updates where the sum of the diffs for each
+`data` in the set is preserved, and every update has a time `t` such that `f
+<=t t`. We write this as `compact(updates, f)`. Since the diffs for each data
+are unchanged, and updates commute, `apply(c, updates) = apply(c,
+compact(updates, f))` for any collection version `c`.
````
I missed this, and we have talked since then, so possibly we should resync! If the intent is something like "copy/paste in the compaction lemmas / intuition" that works for me with some editing. If it is trying to get at a different point, I missed it!

ruchirK reviewed (commented):
I just looked at the commit that added the description of the compaction logic! That looked great and helped me understand this way better than I ever had previously so thank you :)

I had two small suggestions that might make it even more clearer:

1. Would it be possible to somehow indicate the join of `pink dot` with the two elements in `since` visually somehow? You could maybe write `join` or something near those dots maybe?
2. It took me a couple of readings to understand that the goal here is roughly to get rid of as many of the pink dots at grey times, while ensuring that we keep all of the pink dots at blue times, understanding that pink-ness propagates through the arrows. Given that I can sort of visually see that if I removed any more of the pink dots from compact-2.jpg, some of the pink dots at blue times would go away too. I'm not sure if that thought can be cleaned up and useful for the doc, or if its just a rando shower thought type of thing - will leave that up to you.

aphyr commented:
> Would it be possible to somehow indicate the join of pink dot with the two elements in since visually somehow? You could maybe write join or something near those dots maybe?

I've added a new diagram to show the join-then-meet process visually!

> It took me a couple of readings to understand that the goal here is roughly to get rid of as many of the pink dots at grey times, while ensuring that we keep all of the pink dots at blue times, understanding that pink-ness propagates through the arrows. Given that I can sort of visually see that if I removed any more of the pink dots from compact-2.jpg, some of the pink dots at blue times would go away too. I'm not sure if that thought can be cleaned up and useful for the doc, or if its just a rando shower thought type of thing - will leave that up to you.

Yeah! I tried a little rephrasing of the "optimal" discussion to hint at this--I already phrase it a few different ways so I'm not sure if I should include this way too, but open to it!

frankmcsherry commented on doc/developer/platform/formalism.md:
```diff
@@ -13,57 +13,465 @@ It does not intend to describe the **current behavior** of the system.
+
+A *time-varying collection* (TVC) represents the entire history of a collection
+which changes over time. At any time, it has an effective value: a collection
+version. Each TVC is associated with a single timeline, which governs which
```
Yes this is possible. For example, a timeline might be "logical replication from a postgres instance", and one could have multiple postgres instances, each corresponding to different timelines. Their times will be the same set, their meet and join the same operations, but they are distinct timelines whose times should not be related.

umanwizard replied:
In this case, IMO we should add a note to that effect in the definition of timelines, as the language currently there was interpreted (at least by me) to imply otherwise.

aphyr commented:
> In this case, IMO we should add a note to that effect in the definition of timelines, as the language currently there was interpreted (at least by me) to imply otherwise.

I've added an ID to timelines to help with this. :-)

umanwizard commented on doc/developer/platform/formalism.md:
```diff
@@ -13,57 +13,466 @@ It does not intend to describe the **current behavior** of the system.
+=> cv2`. `cv2` is exactly `cv1`, but with `multiplicity(cv2, data) =
+multiplicity(cv1, data) + diff`.
+
+We can also apply a countable set of updates to a collection version by
```
I'm not sure how countably infinite sets of updates can work, unless you restrict the subset of updates at each piece of data to be finite.

For example, consider the set of updates [t0, "SomeData", n] for all positive integers n. The multiplicity of "SomeData" in the result of applying these to any collection version will grow without bound, but our definitions don't allow multiplicities to be infinite. (Or, even if they do, let the diffs in that set be n * (-1) ^ n ).

aphyr replied:
Yeah, infinite multiplicities are a reasonable thing--Reisig defines them in Z. I just said "countable" here because I didn't want to get into whatever the weird Mobius transform pseudointegral was--that math is beyond me. In practice I don't think this matters: we're dealing with computers with finite space and time.

aphyr commented on doc/developer/platform/formalism.md:
```diff
@@ -13,57 +13,466 @@ It does not intend to describe the **current behavior** of the system.
+
+## Collection Versions
+
+A *collection version* is a [multiset](https://en.wikipedia.org/wiki/Multiset)
```
> The definition from that Wikipedia article doesn't match the one you gave here, since it requires that the multiplicity of elements be nonnegative. I don't think there is a generalized term for what you're describing here; a bit of searching suggests that generalized set and hybrid multiset have been used.

Multisets with negative multiplicities have a fairly established track record! Hailperin 1976 calls them "signed multisets", as opposed to "unsigned multisets". Loeb, as you note, calls them "hybrid multisets". "Generalized Set" is from Whitney (1933). Rado (1975) used signed cardinals for multisets, and Reisig (1985) defines multisets as functions to Z. I'll cite Reisig here; it seems simplest.

> Another thing that comes to mind here is that we haven't defined "type" anywhere, but we refer to that concept here.

If you'd like to suggest language I'd love to hear it! The formalism originally talked about types but didn't really define them, and I wasn't quite sure what to say. I'm not sure "all of the collection versions of a given collection have the same domain" helps a whole lot, because we can always define that domain as the union of the individual collection version domains, and it also doesn't constrain the records *within* a single collection version. I kind of got the sense that the original formalism was grasping towards something (e.g.) like an SQL schema such that each piece of data represented a row with well-defined columns, but perhaps this is a question for @frankmcsherry and a future PR?

aphyr commented on doc/developer/platform/formalism.md:
````diff
@@ -13,57 +13,466 @@ It does not intend to describe the **current behavior** of the system.
+```
+diff(t, cv1, cv2) = { u = [data, t, diff] |
+                      (data in cv1 or data in cv2) and
+                      (diff = multiplicity(cv2, data) - multiplicity(cv1, data))
````
Ah, yes, thank you!
